### PR TITLE
Refactor: Use more smart pointers & reduce unnecessary copies

### PR DESCRIPTION
**Pull request by harrynull**

### CODE REVIEW

File: src/ccaligner.cpp [hunk 1]
```diff
@@ -40,7 +40,7 @@ void printFooter()
     std::cout<<"https://github.com/saurabhshri/CCAligner/issues\n";
 }
 
-CCAligner::CCAligner(Params * parameters)
+CCAligner::CCAligner(std::shared_ptr<Params> parameters)
```
Comment (Rakete1111) [inline src/ccaligner.cpp]: Why does `parameters` have to be a shared pointer? There is no need for `CCAligner` to take (shared) ownership of it. A raw pointer is totally fine here.

**Comment by saurabhshri** [comment]
@harrynull the build is broken..

File: src/ccaligner.cpp [hunk 2]
```diff
@@ -66,20 +66,13 @@ int CCAligner::initAligner()
...
-
-}
-
 int main(int argc, char *argv[])
 {
     printHeader("0.03 Alpha [Shubham]");
 
-    Params parameters;
-    parameters.inputParams(argc,argv);
+    auto parameters = std::unique_ptr<Params>(new Params(argc, argv));
```
Comment (Rakete1111) [inline src/ccaligner.cpp]: Why does this have to be a `std::unique_ptr`?

File: src/ccaligner.h [hunk 3]
```diff
@@ -12,13 +12,13 @@
...
-    Params * _parameters;
+    Params* _parameters;
 
 public:
 
-    CCAligner(Params * parameters);
+    CCAligner(Params* parameters);
     int initAligner();                                  //initialize the aligner
-    ~CCAligner();
+    ~CCAligner() = default;
```
Comment (Rakete1111) [inline src/ccaligner.h]: Not needed, it's implicitly defaulted already.

File: src/ccaligner.h [hunk 4]
```diff
@@ -12,13 +12,13 @@
 
 class CCAligner
 {
-    Params * _parameters;
+    Params* _parameters;
```
Comment (Rakete1111) [inline src/ccaligner.h]: The non-existent style guide says that you need a space between `Params` and `*`. Everywhere else in the code it's like that.

File: src/lib_ccaligner/recognize_using_pocketsphinx.cpp [hunk 5]
```diff
@@ -36,15 +36,15 @@ bool PocketsphinxAligner::processFiles()
     LOG("Audio Filename: %s Subtitle filename: %s", _audioFileName.c_str(), _subtitleFileName.c_str());
 
     std::cout << "Reading and processing subtitles...\n";
-    _subParserFactory = new SubtitleParserFactory(_subtitleFileName);
+    _subParserFactory = std::unique_ptr<SubtitleParserFactory>(new SubtitleParserFactory(_subtitleFileName));
```
Comment (Rakete1111) [inline src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: Why a unique pointer here?

Comment (harrynull) [reply src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: or it has to be raw pointer

Comment (Rakete1111) [reply src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: No, I'd say that it should just be a simple object, no need for heap allocations IMO.

Comment (harrynull) [reply src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: @Rakete1111 Note that it is not in constructor, and SubtitleParserFactory has no default constructor.

File: src/lib_ccaligner/params.h [hunk 6]
```diff
@@ -9,11 +9,12 @@
 
 #include "commons.h"
 
-const std::string currentTime();
+std::string currentTime();
 
 class Params
 {
     std::string localTime;
+    void inputParams(int argc, char *argv[]);
```
Comment (saurabhshri) [inline src/lib_ccaligner/params.h]: Though it definitely looks like a good idea, but unfortunately we can not do this. We need to be as flexible as possible. There are many scenarios in which the parameters need be changed mid-way and then brought into usage. Or may be process user input, modify it, and then pass it explicitly multiple times.

**Comment by saurabhshri** [comment]
@harrynull Thanks a lot for the PR, and the amazing work! :) Also, thanks @Rakete1111 for your comments.

**Pull request merged**